stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,246 @@
+        check_finite("algebra_solver", "initial guess", x(i));
+      for (int i = 0; i < y.size(); i++)
+        check_finite("algebra_solver", "parameter vector", y(i));
+      for (size_t i = 0; i < dat.size(); i++)
+        check_finite("algebra_solver", "continuous data", dat[i]);
+
+      if (relative_tolerance <= 0)
+        invalid_argument("algebra_solver",
+                         "relative_tolerance,", relative_tolerance,
+                         "", ", must be greater than 0");
+      if (function_tolerance <= 0)
+        invalid_argument("algebra_solver",
+                         "function_tolerance,", function_tolerance,
+                         "", ", must be greater than 0");
+      if (max_num_steps <= 0)
+        invalid_argument("algebra_solver",
+                         "max_num_steps,", max_num_steps,
+                         "", ", must be greater than 0");
+
+      // Create functor for algebraic system
+      typedef system_functor<F, double, double> FS;
+      typedef hybrj_functor_solver<FS, F, double, double> FX;
+      FX fx(FS(), f, x, y, dat, dat_int, msgs, true);
+      Eigen::HybridNonLinearSolver<FX> solver(fx);
+
+      // Check dimension unknowns equals dimension of system output
+      int z_size = fx.get_value(x).size();  // FIX ME: do w/o computing fx?
+      if (z_size != x.size()) {
+        std::stringstream msg;
+        msg << ", but should have the same dimension as x "
+            << "(the vector of unknowns), which is: "
+            << x.size();
+        std::string msg_str(msg.str());
+        invalid_argument("algebra_solver", "the ouput of the algebraic system",
+                         z_size, "has dimension = ", msg_str.c_str());
+      }
+
+      // Compute theta_dbl
+      Eigen::VectorXd theta_dbl = x;
+      solver.parameters.xtol = relative_tolerance;
+      solver.parameters.maxfev = max_num_steps;
+      solver.solve(theta_dbl);
+
+      // Check if the max number of steps has been exceeded
+      if (solver.nfev >= max_num_steps)
+        invalid_argument("algebra_solver", "max number of iterations:",
+                         max_num_steps, "", " exceeded.");
+
+      // Check solution is a root
+      Eigen::VectorXd system = fx.get_value(theta_dbl);
+      if (system.stableNorm() > function_tolerance) {
+          std::stringstream msg_index;
+          std::stringstream msg_f;
+          msg_f << " but should be lower than the function tolerance: "
+                << function_tolerance
+                << ". Consider increasing the relative tolerance and the"
+                << " max_num_steps.";
+          std::string msg_str_f(msg_f.str());
+
+          invalid_argument("algebra_solver",
+                           "the norm of the algebraic function is:",
+                           system.stableNorm(), "", msg_str_f.c_str());
+      }
+
+      return theta_dbl;
+     }
+
+    /**
+     * Overload the algebraic system to handle the case where y
```

Comment:
Ok. The description of what the function does was documented in the first declaration of the function. I can add it here again, though this seems a little redundant.